decode/decode.go

```diff
@@ -0,0 +1,196 @@
+package decode
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/Clever/syslogparser/rfc3164"
+)
+
+// reservedFields are automatically set during decoding.
+// no field written by a user (e.g. contained in the Kayvee JSON) should overwrite them.
+var reservedFields = []string{
+	"prefix",
+	"postfix",
+	"Type",
+}
+
+func stringInSlice(s string, slice []string) bool {
+	for _, item := range slice {
+		if s == item {
+			return true
+		}
+	}
+	return false
+}
+
+// remapSyslog3164Keys renames fields to match our expecations from heka's syslog parser
+// see: https://github.com/mozilla-services/heka/blob/278dd3d5961b9b6e47bb7a912b63ce3faaf8d8bd/sandbox/lua/decoders/rsyslog.lua
+var remapSyslog3164Keys = map[string]string{
+	"hostname":  "Hostname",
+	"timestamp": "Timestamp",
+	"tag":       "programname",
+	"content":   "Payload",
+}
+
+// FieldsFromSyslog takes an RSyslog formatted log line and extracts fields from it
+//
+// Supports two log lines formats:
+// - RSYSLOG_TraditionalFileFormat - the "old style" default log file format with low-precision timestamps (RFC3164)
+// - RSYSLOG_FileFormat - a modern-style logfile format similar to TraditionalFileFormat, but with high-precision timestamps and timezone information
+//
+// For more details on Rsylog formats: https://rsyslog-5-8-6-doc.neocities.org/rsyslog_conf_templates.html
+func FieldsFromSyslog(line string) (map[string]interface{}, error) {
```

Comment:
Added benchmarks, and they'll be written as Circle artifacts so we can look over time.

8504fa8